---
title: "Tidy data, the tidyverse and tidying data"
subtitle: "Data Science option of BIO00058M Data Analysis."
author: "Emma Rand"
institute: "University of York, UK"
output:
  xaringan::moon_reader:
    css: [default, ../css_files/emma.css, ../css_files/emma-fonts.css]
    lib_dir: libs
    seal: true
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
    fig_caption: false
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      message = FALSE,	
                      warning = FALSE,
                      fig.width = 4, 
                      fig.height = 4, 
                      fig.retina = 3)
options(htmltools.dir.version = FALSE)
```

```{r style-share-again, echo=FALSE}
xaringanExtra::use_share_again()
xaringanExtra::style_share_again(
  share_buttons = "all")
xaringanExtra::use_clipboard()
xaringanExtra::use_extra_styles(
  hover_code_line = TRUE,         
  mute_unhighlighted_code = TRUE)
```

```{r packages, include=FALSE}
library(RefManageR)
library(kableExtra)
library(tidyverse)
```


```{r, load-refs, include=FALSE, cache=FALSE}
BibOptions(check.entries = FALSE,
           bib.style = "authoryear",
           cite.style = "authoryear",
           style = "markdown",
           hyperlink = TRUE,
           dashed = FALSE,
           longnamesfirst = FALSE,
           max.names = 2)
myBib <- ReadBib("../refs.bib", check = FALSE)
```

<style>
div.blue { background-color:#b0cdef; border-radius: 5px; padding: 20px;}
div.grey { background-color:#d3d3d3; border-radius: 0px; padding: 0px;}
</style>

# Outline

The aim of this section is to strength your understanding of the tidyverse `r Cite(myBib, "Wickham2019-ml")` including the pipe, `%>%`  and some commonly applied data tidying operations.

It includes a case study which demonstrates how to problem solve your way to a solution using a one-step-at-a-time approach.

---
# Outline

You should be able to code along with the examples. When you see the film clapper it is ..

`r emo::ji("clapper")` .. an instruction to do something!!


--

`r emo::ji("clapper")` Go through the workflow for creating a [new RStudio Project](https://3mmarand.github.io/BIO00058M-Data-science-2020/workshops/01_project_organisation.html#Task_1_New_RStudio_Project) with the **`usethis`** `r Cite(myBib, "usethis")`. I suggested including directories `R/`, `data-raw/`, and `data-processed/` 

---
class: inverse

#  What is tidy data?


---
# What is tidy data?

Tidy data adhere to a consistent structure which makes it easier to manipulate, model and visualize them. The structure is defined by:  

1. Each variable has its own column.  
2. Each observation has its own row.  
3. Each value has its own cell.  

--

Closely allied to the relational algebra of relational databases `r Cite(myBib, "Codd1990-th")`.  
Underlies the enforced rectangular formatting in SPSS, STATA and R's dataframe/tibble.  
The term 'tidy data' was popularised by `r Citet(myBib, "Wickham2014-nl")`.   

Note: There may be more than one potential tidy structure.

---
# Tidy format

```{r echo=FALSE}
pigeon <- read_table2("../data-raw/pigeon.txt")
```

Suppose we had just 3 individuals in each of two populations:

.pull-left[
NOT TIDY!
two observations per row
column names are not variables
```{r echo=FALSE}
knitr::kable(pigeon[1:3,], 
             format = "html")
```
]
```{r echo=FALSE}
pigeon2 <- gather(data = pigeon, 
                  key = population, 
                  value = distance)
```

.pull-right[

TIDY!
one observation per row
column names are the two variables
```{r echo=FALSE}
knitr::kable(head(pigeon2[c(1:3,41:43),]), 
             format = "html",
             row.names = FALSE)
```
]


---
class: inverse

#  Tidyverse


---
# Tidyverse

The [tidyverse](https://www.tidyverse.org/)  `r Cite(myBib, "R-tidyverse")` is both a paradigm for coding in R and a metapackage (a collection of packages).

Contributors describe it as "an opinionated" collection of R packages designed for data science. 

--

The key feature is that all **`tidyverse`** packages share an underlying design philosophy, grammar, and data structures.  

This means:
-  they work well together
-  learning a new tidyverse packages is quick
-  data work is more efficient.

---
# Tidyverse

**`tidyverse`** packages have a reputation for making code which is easy for humans to read and to write, and for enabling tools to be connected into reproducible workflows.

The R Views [article by Joe Rickert](https://rviews.rstudio.com/2017/06/08/what-is-the-tidyverse/) gives a good overview.

---
# Tidyverse

The are many other extremely useful packages that are not part of the tidyverse but which also use a common design across packages. 

An example is the [Bioconductor Project](https://www.bioconductor.org/).

--

However, Bioconductor packages that take a tidyverse approach are beginning to accumulate. For example:

* **`tidybulk`** `r Cite(myBib, "R-tidybulk")`. See also .font70[https://stemangiola.github.io/rpharma2020_tidytranscriptomics/articles/tidytranscriptomics.html]
* **`biobroom`** `r Cite(myBib, "R-biobroom")`



---
# Tidyverse

You should already have `tidyverse` installed. Packages need installing only once (unless you wish to update them) but must be loaded every session.

`r emo::ji("clapper")` Load the core tidyverse packages with:
```{r eval=FALSE}
library(tidyverse)
```

--
[Core packages](https://www.tidyverse.org/packages/) are: **`ggplot2`**, **`dplyr`**, **`tidyr`**, **`readr`**, **`purr`**, **`tibble`**, **`stringr`** and **`forcats`**

--

Actually, **`ggplot2`** predates the tidyverse which is why it uses `+` to link functions together. It does use the same sort of grammar and works very well within the tidyverse.

---
# Tidyverse

The tidyverse also includes many other packages with more specialised usage. They are not loaded automatically with `library(tidyverse)`and need their own `library()` calls. 

--

Examples are `readxl`, `haven`, `rvest` and `lubridate.`


---
class: inverse

#  The pipe %>%

---
# The pipe %>%

The pipe `%>%` operator from the `magrittr` package `r Cite(myBib, "magrittr")` is loaded with core tidyverse. 


It is the feature that allows us to connect tools together in a readable way.

It can improve code readability by:

--

* structuring sequences of data operations left-to-right or top-to-bottom (as opposed to from the inside and out),

--

* minimizing the need for intermediates, 

--

* making it easy to add steps anywhere in the sequence of operations.

---
# The pipe %>%

The pipe means that instead of using: 
```{r eval=FALSE}
function(object) 
```

We can use 
```{r eval=FALSE}
object %>% function()
```

--

This is useful when you have multiple functions to apply.
---
# The pipe %>%

Instead of:

```{r eval=FALSE}
function2(function1(object)) 
```

We can use 
```{r eval=FALSE}
object %>% function1()  %>% function2()
```

---
# The pipe %>%

As a simple example, suppose we want to apply a log-squareroot transformation<sup>1</sup> to some proportion data. 

.font70[
.footnote[1. a transformation commonly applied to proportion data to make it less platykurtic $x^{t} = log(\sqrt{x})$
]
]

--

`r emo::ji("clapper")` Generate a random sample of ten proportions to work with:

```{r}
nums <- sample((1:100)/100, size = 10, replace = FALSE)
```

--

Two ways we *could* apply the log-squareroot transformation are to:

1. nest the squareroot and log functions  
2. create intermediate variables  

---
# The pipe %>%

`r emo::ji("clapper")` Nest the `sqrt()` and `log()` functions:

```{r}
tnums <- log(sqrt(nums))
```

--

Code must read from inside to outside. 

Increasingly difficult to read as number of functions increases. 

Also makes simple debugging harder.

---
# The pipe %>%

`r emo::ji("clapper")` Create intermediate variables:

```{r}
sqrtnums <- sqrt(nums)
tnums <- log(sqrtnums)
```

--

Easier to read than nesting.

But you have extra variables you don't need and which become increasingly difficult to name appropriately creating code and workspace clutter.

---
# The pipe %>%

Using the pipe avoids these by taking the output of one operation as the input of the next. 

The pipe has long been used by Unix operating systems (where the pipe operator is `|`). 

The R pipe operator is `%>%` <sup>1</sup>


.font70[
.footnote[1. a pipe recently became available in base R too (with R version 4.1.0). Here the operator is `|>`
]
]

--

`r emo::ji("clapper")` The keyboard short cut is ctrl-shift-M.

---
# The pipe %>%

`r emo::ji("clapper")` Use pipes to code the functions in sequence:
```{r}
tnums <- nums %>% 
  sqrt() %>% 
  log()
```

--

This can be read as: take `nums` *and then* squareroot it *and then* log it.

---
# The pipe %>%

More explicitly, this is:

```{r}
tnums <- nums %>% 
  sqrt(.) %>% 
  log(.)
```

Where `.` stands for the object being passed in. 

In most cases, you don't need to include the `.`.

Occasionally you do have to, for example when arguments are optional or there is ambiguity over which argument is meant.

---
class: inverse

#  Data tidying tasks

---
# Data tidying tasks

Tidying data includes reshaping it in to 'tidy' format but also other tasks such as:

* renaming variables for consistency  
* recoding variables  
* cleaning content for consistency with respect to valid values, missing values and NA  

--

`r emo::ji("eyes")` Key point!

* Keep the raw data exactly as it came to you and do not alter/edit.
* Script and document all tidying tasks.

---
class: inverse
# Converting "wide" to "long"

---
# Converting "wide" to "long"

Data commonly need to be reshaped from a format with more than one observation per row.

--

The data given in [biomass.txt](../data-raw/biomass.txt) are taken from an experiment in which the insect pest biomass (g) was measured on plots sprayed  with water (control) or one of five different insecticides. 

Also in the data file are variables indicating the replicate number and the identity of the tray in which the plant was grown.

---
# Converting "wide" to "long"

`r emo::ji("clapper")` Save a copy of this file to your `data-raw` folder and read it in.

```{r}
file <- "../data-raw/biomass.txt"
biomass <- read_table(file)

```

`read_table()` is a function from the tidyverse's **`readr`** package. It works a lot like base R's `read.table()` but with some useful defaults (e.g., is header assumed) and extra functionality (e.g.,it's faster).

tidyverse import functions treat strings as character variables by default, not factors. Until R 4.x, `read.table()` turned strings into factors. This used to be an important reason for me to use `read_table()`.

Whilst analysis and visualisation often require factor variables, any processing of strings is made much easier if they are characters.

---
# Converting "wide" to "long"

`r emo::ji("clapper")` View the dataframe.

.scroll-output-height[
.code70[
```{r echo=FALSE}
biomass
```
]
]


A tibble is the tidyverse's dataframe. In fact, tibbles are dataframes as well as tibbles.

---
# Converting "wide" to "long"

These data are in "wide" format and can be converting to "long" format using the `tidyr` package function `pivot_longer()`. 

--

`pivot_longer()` collects the values from specified columns (`cols`) into a single column (`values_to`) and creates a column to indicate the group (`names_to`). 

--

We want to gather the first 6 columns but the `rep_tray` column contents should be repeated.


---
# Converting "wide" to "long"

`r emo::ji("clapper")` Gather all the columns of biomass except `rep_tray`:

```{r}
biomass2 <- biomass %>% 
  pivot_longer(names_to = "spray", 
         values_to = "mass",
         cols = -rep_tray)
```

The values will be in a column called `mass`, the treatments in a column called `spray`.

---
# Converting "wide" to "long"

`r emo::ji("clapper")` View the dataframe.

.scroll-output-height[
.code70[
```{r echo=FALSE}
biomass2
```
]
]

---
# Converting "wide" to "long"

`r kableExtra::text_spec("Extra exercise:", background = "#25496b", color = "#fdf9f6", bold = TRUE)` Write this dataframe to file to your `processed_data` folder. This was covered in the [Project Organisation workshop](https://3mmarand.github.io/BIO00058M-Data-science-2020/workshops/01_project_organisation.html)


```{r echo=FALSE}
file <- "../data-processed/biomass2-tidyverse.txt"
write_delim(biomass2, 
            file = file)
```

See the result: [biomass2-tidyverse.txt](../data-processed/biomass2-tidyverse.txt)

---
class: inverse

# Splitting column contents
---
# Splitting column contents

We sometimes have single columns which contain more than one type of encoded information. 

--

For example, if a column contains a full species name you may wish to separate the genus and species into separate columns. This allows you to perform a by-genus analysis.

--

Another example arises when you read in multiple files with names that encode a date, experiment or treatment. We  can add the file name to a column and then split the name into columns for the date, experiment or treatment.

---
# Splitting column contents

For the `biomass2` data we might wish to separate the replicate number from the tray identity and put them in two separate columns. 

--


We can do this with a 'regular expression' or [regex](https://en.wikipedia.org/wiki/Regular_expression). A regex defines a pattern for matching text. 

--

It's a big topic and there are many tutorials. I remember a few bits and google "how to match ... regex". 

[A quick reference](https://www.rexegg.com/regex-quickstart.html)

---
# Splitting column contents

The `extract()` function from the `tidyr` helps us achieve this.

We give:
-  the names of the new columns we want to create
-  the patterns matching the part of the `rep_tray` value we want to go in each column.

---
# Splitting column contents

`r emo::ji("clapper")` Extract parts of `rep_tray` and put in new columns `replicate_number`, `tray_id`:

```{r}
biomass3 <- biomass2 %>% 
  extract(rep_tray, 
          c("replicate_number", "tray_id"),
          "([0-9]{1,2})\\_([a-z])")
```

--
.font80[
-  The patterns to save into columns are inside `( )`.  
-  The pattern going into `replicate_number`, `[0-9]{1,2}`, means 1 or 2 numbers.  
-  The pattern going into `tray_id`, `[a-z]` means one lowercase letter.
-  the bit between the two `( )`, `\_` is a pattern that matches what is in `rep_tray` but is not to be saved. 
]
 

---
class: inverse
# Case study

---
# Case study

## Overivew

You will now work through an example of some real data from [The Genever Group](https://www.geneverlab.info/). The arrangement and format of these data are typical of many protein and gene expression datasets so the processing is representative of that needed in a variety of situations.

--

The data are mass spectrometry data of the soluble protein fraction from five immortalised mesenchymal stromal cell (MSC) lines. 

The data are normalised protein abundances. Each row is a protein.

---
# Case study

`r emo::ji("clapper")` Save a copy of [Y101_Y102_Y201_Y202_Y101-5.csv](../data-raw/Y101_Y102_Y201_Y202_Y101-5.csv) file to your `data-raw` folder.

`r emo::ji("clapper")` You may wish to view it in excel while reading the information on the next page.

---
# Case study

**Data description**

The cells lines are Y101, Y102, Y201, Y202 and Y101.5 and there are three replicates for each cell line arranged in columns. Also in the file are columns for:
.font70[
-  the protein accession
-  the number of peptides used to identify the protein
-  the number of unique peptides used to identify the protein
-  a measure of confidence in that identification
-  the maximum fold change between the mean abundances of two cell lines (i.e., highest mean / lowest mean)
-  a *p* value for a comparison test between the highest mean and lowest mean
-  a *q* value (corrected *p* value) 
-  a measure of the power of that test
-  the cell line with the highest mean
-  the cell line with the lowest mean
-  the protein mass
-  whether at least two peptides were detected for a protein.
]
---
# Case study

**Data Import**

Column names are spread over three rows but are primarily in the third row.

--

We can read in from the third row by skipping the first two. We can also use the `clean()` function from the **`janitor`** package to improve the column names.

---
# Case study

**Data Import**

`r emo::ji("clapper")` Read in the file using `read_csv()` from the **`tidyverse`**'s **`readr`** package.
```{r}
# define file name
filesol <- "../data-raw/Y101_Y102_Y201_Y202_Y101-5.csv"

# skip first two lines
sol <- read_csv(filesol, skip = 2) %>% 
  janitor::clean_names()

```

---
# Case study

**Data Import**

`r emo::ji("eyes")` the `::` notation gives you access to a package's functions without first using the `library()` command. 

This is useful when you only want to use one function from a package, or when you need to specify which package the function is in. 

---
# Case study

**Filtering rows**

The cells were grown on bovine serum and any bovine proteins need to be filtered out.

--

If a protein identify has been determined by fewer than two peptides we are not confident in that identification. Such proteins should also be filtered out. This is common practice for  proteomic data. 

---
# Case study

**Filtering rows**

**`dplyr`** (yep, a tidyverse package) has the `filter()` function. It is easier to use than selection using `dataframe[condition1 & condition2]`

`r emo::ji("clapper")` Keep rows of human proteins identified by more than one peptide:
```{r}
sol <- sol %>% 
  filter(str_detect(description, "OS=Homo sapiens")) %>% 
  filter(x1pep == "x")
```

---
# Case study

`r emo::ji("eyes")` `str_detect(string, pattern)` returns a logical vector according to whether 'pattern' is found in 'string'.

`r emo::ji("eyes")` Notice that we have applied `filter()` twice using the pipe.

---
# Case study

**Processing cells contents**

It would be useful to extract the genename from the `description` and put it in a column of its own.

One entry from the `description` column looks like this:

.scroll-output-width[
.code60[
```{r}
sol$description[1]
```
]
]

--

The genename is after `GN=`. We need to extract the part of the string that follows `GN=` and put it in a new column. We need to do that for every row in of the dataframe.

---
# Case study

**Processing cells contents**

You can problem-solve your way to the result by breaking the steps down. Start with *one* value and perform operations one at a time until you've worked out the pipeline. Then implement on the entire column.

The pipe makes it especially easy to break problems down like this.

---
# Case study

**Processing cells contents**

One step at a time on one value.

We need to:
-  get one value from `description` to practice with
-  find `GN=` in the description
-  extract the string that starts `GN=`
-  remove the `GN=` part of that that string


---
# Case study

**Processing cells contents**

One step at a time on one value.


`r emo::ji("clapper")` Extract the first value of the `description` to work with:
```{r}
one_description <- sol$description[1]
```

--

We could use `sol$description[2]` or `sol$description[100]` or any other row


---
# Case study

**Processing cells contents**

One step at a time on one value.

`r emo::ji("clapper")` Extract the string that starts `GN=` using a regex:
```{r}
str_extract(one_description,"GN=[^\\s]+")
```

Explanation on next slide.

---
# Case study

**Processing cells contents**

One step at a time on one value.


* `[ ]` means some characters
* `^` means 'not' when inside `[ ]`
* `\s` means white space  
* the `\` before is an escape character to indicate that the next character, `\` should not be taken literally (because it's part of `\s`)  
* `+` means one or more  

--

So `GN=[^\\s]+` means `GN=` followed by one or more characters that are not white space. This means the pattern stops matching at the first white space after "GN=".

---
# Case study

**Processing cells contents**

One step at a time on one value.

We're close. Now we will drop the `GN=` part by replacing it with nothing:

`r emo::ji("clapper")` Add replacing `GN=` with an empty string, `""`, to the pipeline:
```{r}
str_extract(one_description, "GN=[^\\s]+") %>% 
  str_replace("GN=", "")
```

--
.font140[
`r emo::ji("party")`
]
---
# Case study

**Processing cells contents**

Creating a new column

Now we know how to get the result for one value, we need to apply the same process to the whole column

--

`mutate()` is the **`dplyr`** function that adds new variables and preserves existing ones. It takes `name` = `value` pairs of expressions where:

* `name` is the name for the new variable and 
* `value` is the value it takes. This is usually an expression. 

---
# Case study

**Processing cells contents**

Creating a new column

`r emo::ji("clapper")` Add a variable `genename` which contains the processed string from the `description` variable:
```{r}
sol <- sol %>%
  mutate(genename =  str_extract(description,"GN=[^\\s]+") %>% 
           str_replace("GN=", ""))

```

---
# Case study

You will continue to work on this case study in the workshop.

---
# Summary

.font80[

-  **`tidyverse`** is a collection of packages with a common design; many excellent packages are not part of it  
-  `library(tidyverse)` loads the core packages; other **`tidyverse`** packages need their own `library()` call  
-  the pipe `%>%` is key to connecting **`tidyverse`** tools together to create highly readable code  
-  reshaping data from wide to long format is common: `pivot_longer()`  
-  splitting cell contents is common: `extract()` 
-  regular expressions allow you to match text patterns; expect to have to google and do a lot of trial and error  
-  `clean_names()` is well useful!
-  `::` gives you access to a package's functions without using `library()`  
-  use particular rows `filter()`  
  

]

---
# Reading

## Strongly recommended

-  [article by Joe Rickert](https://rviews.rstudio.com/2017/06/08/what-is-the-tidyverse/) gives a good overview.
-  Tidy Data `r Cite(myBib, "Wickham2014-nl")` sections 1 and 2

## Further
-  Welcome to the Tidyverse `r Cite(myBib, "Wickham2019-ml")` 
-  Tidy Data `r Cite(myBib, "Wickham2014-nl")` sections 3 - 6
-  R for Data Science `r Cite(myBib, "Wickham:2017:RDS:3086927")` Chapter 12.1, to 12.3 and Chapter 18


---
# References
.font50[
.footnote[
Slides made with xaringan `r Cite(myBib, "xaringan")`,  xaringanExtra `r Cite(myBib, "xaringanExtra")`, xaringanthemer `r Cite(myBib, "xaringanthemer")` and RefManageR `r Cite(myBib, "RefManageR")` 

]

```{r refs, echo=FALSE, results="asis"}
PrintBibliography(myBib)
```

]


---

## Emma Rand <br> [emma.rand@york.ac.uk](mailto:emma.rand@york.ac.uk) <br> Twitter: [@er13_r](https://twitter.com/er13_r) <br> GitHub: [3mmaRand](https://github.com/3mmaRand)  <br> blog: https://buzzrbeeline.blog/
<br>
<a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br /><span xmlns:dct="http://purl.org/dc/terms/" property="dct:title">Data Science strand of BIO00058M</span> by <span xmlns:cc="http://creativecommons.org/ns#" property="cc:attributionName">Emma Rand</span> is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.
